### PR TITLE
[VLM] Support ViT Piecewise CUDA Graph for Qwen3-VL

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

This PR is to enable ViT Piecewise CUDA Graph for Qwen3-VL. 
Building logic upon ViTCudaGraphRunner to support both Qwen2.5-VL and Qwen3-VL.
TP>1 is supported.

Benchmark show 8xH20 Qwen3-VL-8B-Instruct TP=4
TTFT 1384.53ms --> 1120.68ms

Meanwhile fixed a bug that torch-symm-mem is not enabled for outplace allreduce. It gains about 4% e2e improvement over NCCL. (As custom all reduce has not supported CUDA Graph yet, we have to disable-custom-all-reduce. Thanks to torch-symm-mem, which gives extra 4% speedup over legacy NCCL TP. But the TTFT comparation in this PR is torch-symm-mem vs torch-symm-mem, the only difference is enable/disable ViT Piecewise CUDA Graph)

The sweet spot for this feature is each rank's compute is relatively small, (i.e. TP is 4, image size is not large, no compute bound in prefill) so the kernel launch occupies large portion. ViT in PCG can save this time cost.

## Detailed Design

image

## Accuracy Tests

The update in test/manual/nightly/test_vlms_vit_cuda_graph.py case covers the accuracy.

## Benchmarking and Profiling

8xH20 Qwen3-VL-8B-Instruct TP=4
TTFT 1384.53ms --> 1120.68ms

PR:
Server:
```
SGLANG_MM_FEATURE_CACHE_MB=4096 \
SGLANG_USE_CUDA_IPC_TRANSPORT=1 \
SGLANG_VLM_CACHE_SIZE_MB=0 \
SGLANG_VIT_ENABLE_CUDA_GRAPH=1 \
python3 -m sglang.launch_server --host 127.0.0.1 --mem-fraction-static 0.7 --port 30000 --max-running-requests 64 --chunked-prefill-size 8192 --attention-backend fa3 --mm-attention-backend fa3 --enable-multimodal --model /home/admin/Qwen3-VL-8B-Instruct --disable-radix-cache --piecewise-cuda-graph-max-tokens 4096 --enable-piecewise-cuda-graph --piecewise-cuda-graph-compiler eager --tp-size 4 --disable-custom-all-reduce --enable-torch-symm-mem
```

Client:
```
$python3 -m sglang.bench_serving \
  --backend sglang-oai-chat \
  --dataset-name image \
  --num-prompts 256 \
  --apply-chat-template \
  --random-input-len 128 \
  --random-output-len 1 \
  --image-resolution 560x560 \
  --image-format jpeg \
  --image-count 1 \
  --image-content random \
  --random-range-ratio 0.1 \
  --port 30000 \
  --max-concurrency 32 \
  --warmup-requests 5
```

Benchmark Result:
```
============ Serving Benchmark Result ============
Backend:                                 sglang-oai-chat
Traffic request rate:                    inf       
Max request concurrency:                 32        
Successful requests:                     256       
Benchmark duration (s):                  9.27      
Total input tokens:                      103986    
Total input text tokens:                 20530     
Total input vision tokens:               83456     
Total generated tokens:                  256       
Total generated tokens (retokenized):    256       
Request throughput (req/s):              27.61     
Input token throughput (tok/s):          11213.80  
Output token throughput (tok/s):         27.61     
Peak output token throughput (tok/s):    47.00     
Peak concurrent requests:                65        
Total token throughput (tok/s):          11241.41  
Concurrency:                             30.94     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   1120.68   
Median E2E Latency (ms):                 1072.01   
---------------Time to First Token----------------
Mean TTFT (ms):                          1120.67   
Median TTFT (ms):                        1072.00   
P99 TTFT (ms):                           1836.13   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00      
Median TPOT (ms):                        0.00      
P99 TPOT (ms):                           0.00      
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

Baseline:
Server:
```
SGLANG_MM_FEATURE_CACHE_MB=4096 \
SGLANG_USE_CUDA_IPC_TRANSPORT=1 \
SGLANG_VLM_CACHE_SIZE_MB=0 \
python3 -m sglang.launch_server --host 127.0.0.1 --mem-fraction-static 0.7 --port 30000 --max-running-requests 64 --chunked-prefill-size 8192 --attention-backend fa3 --mm-attention-backend fa3 --enable-multimodal --model /home/admin/Qwen3-VL-8B-Instruct --disable-radix-cache --piecewise-cuda-graph-max-tokens 4096 --enable-piecewise-cuda-graph --piecewise-cuda-graph-compiler eager --tp-size 4 --disable-custom-all-reduce --enable-torch-symm-mem
```

Client:
Same as above.

Result:
```
============ Serving Benchmark Result ============
Backend:                                 sglang-oai-chat
Traffic request rate:                    inf       
Max request concurrency:                 32        
Successful requests:                     256       
Benchmark duration (s):                  11.33     
Total input tokens:                      104004    
Total input text tokens:                 20548     
Total input vision tokens:               83456     
Total generated tokens:                  256       
Total generated tokens (retokenized):    252       
Request throughput (req/s):              22.59     
Input token throughput (tok/s):          9178.06   
Output token throughput (tok/s):         22.59     
Peak output token throughput (tok/s):    32.00     
Peak concurrent requests:                64        
Total token throughput (tok/s):          9200.65   
Concurrency:                             31.28     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   1384.53   
Median E2E Latency (ms):                 1351.33   
---------------Time to First Token----------------
Mean TTFT (ms):                          1361.75   
Median TTFT (ms):                        1350.20   
P99 TTFT (ms):                           1982.05   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00      
Median TPOT (ms):                        0.00      
P99 TPOT (ms):                           0.00      
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

**Comment by yuan-luo** [comment]
https://github.com/sgl-project/sglang/issues/15321

File: python/sglang/srt/models/qwen3_vl.py [hunk 1]
```diff
@@ -439,6 +447,9 @@ def forward(
         x: torch.Tensor,
         grid_thw: torch.Tensor,
     ) -> torch.Tensor:
+        if get_bool_env_var("SGLANG_VIT_ENABLE_CUDA_GRAPH"):
+            return self.forward_with_cuda_graph(x, grid_thw)
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/models/qwen3_vl.py:470]: Now this function works in disable-custom-all-reduce for TP>1. So the PR implementation stands.
```
$SGLANG_MM_FEATURE_CACHE_MB=4096 SGLANG_USE_CUDA_IPC_TRANSPORT=1 SGLANG_VLM_CACHE_SIZE_MB=0 SGLANG_VIT_ENABLE_CUDA_GRAPH=1 python3 -m sglang.launch_server --host 127.0.0.1     --mem-fraction-static 0.7     --port 30000     --max-running-requests 64     --chunked-prefill-size 8192     --attention-backend fa3     --mm-attention-backend fa3     --enable-multimodal     --model /home/admin/Qwen3-VL-8B-Instruct     --disable-radix-cache     --piecewise-cuda-graph-max-tokens 128     --enable-piecewise-cuda-graph     --piecewise-cuda-graph-compiler eager --tp-size 2 --disable-custom-all-reduce
```

File: python/sglang/srt/models/qwen3_vl.py [hunk 2]
```diff
@@ -474,6 +485,47 @@ def forward(
...
+            position_embeddings[1].to(x.device, x.dtype),
+        )
+
+        # compute cu_seqlens
+        cu_seqlens = compute_cu_seqlens_from_grid_numpy(grid_thw)
+        if not isinstance(cu_seqlens, torch.Tensor):
+            cu_seqlens = torch.tensor(cu_seqlens, device=x.device, dtype=torch.int32)
+        else:
+            cu_seqlens = cu_seqlens.to(device=x.device, dtype=torch.int32)
+        cu_seqlens = cu_seqlens.contiguous()
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/models/qwen3_vl.py:542]: Prefer to keep intact as compute_cu_seqlens_from_grid_numpy might be refactored as it is a common utils library used by many VLMs.

File: test/manual/nightly/test_vlms_vit_cuda_graph.py [hunk 3]
```diff
@@ -145,7 +146,7 @@ def _run_vlm_mmmu_test(
                     "fa3",
                     "--enable-piecewise-cuda-graph",
                     "--piecewise-cuda-graph-max-tokens",
-                    "8192",
+                    "1024",
```
Comment (BBuf) [inline test/manual/nightly/test_vlms_vit_cuda_graph.py]: Why change `--piecewise-cuda-graph-max-tokens`

**Review by BBuf — APPROVED** [review]
LGTM

File: test/manual/nightly/test_vlms_vit_cuda_graph.py (hunk 3 above)
Comment (yuan-luo) [reply test/manual/nightly/test_vlms_vit_cuda_graph.py]: The initiative is to make CI faster. After thinking twice, should change back, for the token > 1024 is out of being covered by Piecewise CG.
